### DISCUSSION TITLE
EV3Console上に表示されるメッセージについて

### DISCUSSION THREAD

**Discussion by t-hara01** [toppers/users-forum#91] [ET-Robocon]
ETロボコン2022に参加中の者です。

チーム内でなかなか解決できない問題があり
質問させていただきたいです。

【質問】
プログラム実行中にEV3Console上に⇓画像のような表示が出るのはどのような時でしょうか
 投げやりな質問で申し訳ありません。自分たちでは長らく解決できずにこちらに投稿させていただきます。

【使用しているPC・環境】
PC:Windows10
走行体:EV3RT HackEV
開発言語: C/C++

【現状】
自分たちが作成したETロボコン競技用のプログラムを実行する際
実行途中に例外が補足されプログラムが強制終了する時があります。
毎回ではなく、数回に一回という頻度です。
一度強制終了が起こった場合はEV3本体の電源OFF後再起動するまで、異常終了する状態が続きます。

コンソールログを見ると↓画像のように出力されています。
「translation fault(1st level)」の部分で
何かの変換に失敗してプログラムが終了していると考えています。
![スクリーンショット 2022-09-20 175520](https://user-images.githubusercontent.com/51468905/191465051-9e60787b-b966-4a55-86c1-153604f4fd7a.png)

作成中のプログラム内でデータ型を混在させて利用しているのがまずかったと予想を立てて
プログラム修正を行いましたが、改善せずでした。
(int⇒int32_tへ統一など)

**Comment by hiroakitakada** [comment]
Data Abort 例外が起こっているということは，不正な番地にアクセスしているものと思われます。その結果，MMUがアドレス変換例外を起こしています。
EV3RTの出しているメッセージの正確な意味がわかっていないのですが，PCが 0xc06270e4 となっていますので，この番地かその手前の番地に置かれている命令が原因と思われます。この番地に置かれているのがユーザプログラムであれば，原因はユーザプログラムにあると思います。
とりあえず，すぐにわかる範囲で回答させていただきました。

**Reply by hiroakitakada** [reply]
EV3RTの開発者に聞いてみました。

> PCが 0xc06270e4 となっていますので，この番地かその手前の番地に置かれている命令が原因と思われます。

と書きましたが，この番地はEV3にロードされた後の番地です。ロードモジュールの中での番地は，virtual address として表示されている 0x000070e4 になりますので，この番地でどのプログラムが問題かを探ってもらうと良いと思います。

**Reply by t-hara01** [reply]
大会直前で焦っている中だったので教えていただけて、大変助かりました。
ありがとうございます。
これから自分たちのプログラムをもう少し調べてみようと思います。
メモリ番地とユーザプログラムが対応した状態で表示できるようなツールがあるのでしょうか？

**Reply by YixiaoLi** [reply]
「arm-none-eabi-objdump -DS --prefix-addresses app」（appはプログラム名）という命令で、以下のようにDisassembly情報（メモリ番地、関数名、アセンブラ命令）を出力できます。例えば、00000054の番地は、itron_strerrorという関数の先頭で、「lsl r0, r0, #24」というアセンブラ命令が入っていると確認できます。このツールで、エラー情報のvirtual address（00006580と000070e4）が入っている関数を特定可能で、原因の調査に使えると思います。
![Screenshot](https://user-images.githubusercontent.com/11767396/191700280-4550fedc-3ddb-40c8-81d1-0a9ccb50c87e.png)

**Reply by t-hara01** [reply]
@YixiaoLi このようなツールがあるのですね！!
原因箇所特定に希望が見えてきました！ありがとうございます。
